test/functional/owners_controller_test.rb

```diff
@@ -0,0 +1,213 @@
+        end
+        should "set success notice flash" do
+          expected_notice = "Owner added successfully. A confirmation mail has been sent to #{@new_owner.handle}'s email"
+          assert_equal expected_notice, flash[:notice]
+        end
+      end
+
+      context "with incorrect params" do
+        should "show error message if user not found" do
+          post :create, params: { handle: "no_user", rubygem_id: @rubygem.name }
+          expected_alert = "User must exist"
+          assert_equal expected_alert, flash[:alert]
+        end
+
+        should "show error message if ownership exists" do
+          @new_owner = create(:user)
+          create(:ownership, rubygem: @rubygem, user: @new_owner)
+          post :create, params: { handle: @new_owner.handle, rubygem_id: @rubygem.name }
+          expected_alert = "User has already been taken"
+          assert_equal expected_alert, flash[:alert]
+        end
+      end
+    end
+
+    context "on DELETE to owners" do
+      setup do
+        @owner = create(:user)
+        @ownership = create(:ownership, rubygem: @rubygem, user: @owner)
+      end
+
+      context "remove user as gem owner" do
+        setup do
+          delete :destroy, params: { rubygem_id: @rubygem.name, handle: @owner.display_id }
+        end
+        should redirect_to("ownership index") { rubygem_owners_path(@rubygem) }
+        should "remove the ownership record" do
+          refute @rubygem.owners_including_unconfirmed.include?(@owner)
+        end
+      end
+
+      context "not remove last confirmed owner" do
+        setup do
+          @ownership.destroy
+          @last_ownership = @rubygem.ownerships.last
+          delete :destroy, params: { rubygem_id: @rubygem.name, handle: @last_ownership.user.display_id }
+        end
+        should redirect_to("ownership index") { rubygem_owners_path(@rubygem) }
+        should "remove the ownership record" do
+          assert @rubygem.owners_including_unconfirmed.include?(@last_ownership.user)
+        end
+        should "should flash error" do
+          assert_equal "Owner cannot be removed!", flash[:alert]
+        end
+      end
+    end
+
+    context "on GET to resend confirmation" do
+      setup do
+        @new_owner = create(:user)
+        @ownership = create(:ownership, :unconfirmed, rubygem: @rubygem, user: @new_owner)
+        sign_in_as(@new_owner)
+      end
+      context "with correct params" do
+        setup do
+          get :resend_confirmation, params: { rubygem_id: @rubygem.name, handle: @new_owner.display_id }
+        end
+
+        should redirect_to("rubygem show") { rubygem_path(@rubygem) }
+        should "set success notice flash" do
+          success_flash = "A confirmation mail has been re-sent to #{@new_owner.handle}'s email"
+          assert_equal success_flash, flash[:notice]
+        end
+      end
+
+      context "with incorrect params" do
```

Comment:
It is fine if you want to group all failures in a `with incorrect params` context. However, the following three should blocks need their own context.
```
context "gem was not found" do
   should "show 404 error page" do
  end
end
```